### ISSUE TITLE
srcref jumping with dangling srcfiles

### ISSUE THREAD

**Issue by lionel-** [emacs-ess/ESS#553]
When a package is installed from a transient location (as would be the case when installing from CRAN or github), the source files recorded in the srcrefs no longer exist and this prevents xref jumping.

As it turns out, the srcrefs contain a full copy of their original file:

```r
srcref <- attr(fn, "srcref")
srcfile <- attr(srcref, "srcfile")
srcfile$lines
```

We could first check if `srcfile$filename` points to an actual file. If not the case, we'd walk over all functions in the namespace to collect all `srcfile` environments and then regenerate a copy of the source in a temporary location. Further jumps would redirect there.

What do you think @vspinu @atheriel ?

**Comment by lionel-** [comment]
Also Tomas is working on fixing the overhead caused by installing packages with srcrefs. His goal is to make `options(keep.source = TRUE)` the default which would make the mechanism above immediately useful to all users.

**Comment by vspinu** [comment]
Regenerate full namespace is an overkill (plus, sources of some functions can change and keeping a cache asks for trouble). 

Second step should be (and is already there) to look into `ess-r-package-library-paths` for the package with the same name, if not found a temporary copy of the source should be created through dump-to-file functionality (not yet implemented; was my task but I am slacking :/).

**Comment by lionel-** [comment]
I don't think it's an overkill because it's helpful to navigate files when you're exploring a code base. I think we'll want this functionality to work for all functions and not just those installed locally, especially if srcrefs get installed by default.

Cache invalidation is easy: srcfile envs are time-stamped.

**Comment by lionel-** [comment]
Only the `src/` folder would be missing, all the rest would be there (the whole `R/` folder, `NAMESPACE`, `DESCRIPTION`). I'd say it's more than a half-measure and would bring the source jumping experience close to emacs lisp.

**Comment by lionel-** [comment]
I really want this to just work. Maybe we should pull the whole CRAN package as you suggest but that would result in a noticeable delay.

**Comment by vspinu** [comment]
> the whole R/ folder, 

Are you saying that the whole R file could be retrived 100% identical with the original source? I didn't know that. In that case, it might be worth a try indeed.

**Comment by lionel-** [comment]
Yes the srcrefs of a function have lines number + a reference to a sourcefile environment which contains the whole file (in the `lines` field). I didn't know about this either, though in retrospect that should have been obvious since roxygen is based on srcrefs.

**Comment by vspinu** [comment]
That `lines` objects is extremely weird. There are not even new lines over there. 

When I said sources could be modified, I meant when we source into a package environment, we need to pick the right (modified) source. We should use the src refs generated by our eval functions and merge them into original src sources of the pacakge. This way we are back to composing sources from individual pieces. Then, you have to check for staled sources of all functions every time the end user xrefs.

**Comment by atheriel** [comment]
I encountered this inconsistency when I worked on the original `srcref` stuff, but I was never able to work out the reason. It was not clear to me what was going on in the R source itself, either -- maybe it's documented somewhere?

And FWIW, I don't feel that strongly about the feature one way or the other. It sounds attractive for the times you want to read the source in its original context, but I'm not sure this is the way to get there. As a user I wouldn't mind being prompted to download the source of a package I was interested in before falling back on something like dump-to-file. Usually I find myself popping open GitHub and using the search functionality, but it would certainly be nicer to navigate in Emacs.

**Comment by lionel-** [comment]
> it would certainly be nicer to navigate in Emacs.

In my mind there is no comparison. With generalised srcrefs you could go from functions to functions across namespace boundaries transparently, with jump history etc. With increasingly complex dep trees this would be game changing.

**Comment by vspinu** [comment]
Having navigation in a context is indeed appealing. Similar functions and helper function are usually grouped within a single file and having immediate access to all those functions seems like an important benefit. Also imenu based tooling would work there.

The real issue is the correct and robust implementation, but that's the part for @lionel- to figure out :P
